<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>References | Classification and Clustering: Case Study with CRC Gene Expression Data</title>
  <meta name="description" content="Classification and clustering methodes are explored using colorectal data from GEO website" />
  <meta name="generator" content="bookdown 0.20.6 and GitBook 2.6.7" />

  <meta property="og:title" content="References | Classification and Clustering: Case Study with CRC Gene Expression Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Classification and clustering methodes are explored using colorectal data from GEO website" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="References | Classification and Clustering: Case Study with CRC Gene Expression Data" />
  
  <meta name="twitter:description" content="Classification and clustering methodes are explored using colorectal data from GEO website" />
  

<meta name="author" content="Francois Collin" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="summary.html"/>

<script src="libs/header-attrs-2.4/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.5.2/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.15/datatables.js"></script>
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.0.1/js/crosstalk.min.js"></script>
<script type="text/javascript">
// source:https://stackoverflow.com/questions/45360998/code-folding-in-bookdown
// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Genomic Scale Classification and Clustering</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preamble</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#outline"><i class="fa fa-check"></i><b>1.1</b> Outline</a></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#data-preprocessing"><i class="fa fa-check"></i><b>1.2</b> Data Preprocessing</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#affymetrix-probe-sets-vs-genes"><i class="fa fa-check"></i><b>1.3</b> Affymetrix Probe Sets vs Genes</a></li>
<li class="chapter" data-level="1.4" data-path="intro.html"><a href="intro.html#building-a-classifier"><i class="fa fa-check"></i><b>1.4</b> Building a Classifier</a></li>
<li class="chapter" data-level="1.5" data-path="intro.html"><a href="intro.html#classifier-assessment"><i class="fa fa-check"></i><b>1.5</b> Classifier Assessment</a></li>
<li class="chapter" data-level="1.6" data-path="intro.html"><a href="intro.html#subclass-discovery"><i class="fa fa-check"></i><b>1.6</b> Subclass Discovery</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="load-data.html"><a href="load-data.html"><i class="fa fa-check"></i><b>2</b> Load Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="load-data.html"><a href="load-data.html#download-datasets"><i class="fa fa-check"></i><b>2.1</b> Download Datasets</a></li>
<li class="chapter" data-level="2.2" data-path="load-data.html"><a href="load-data.html#peek-into-datasets"><i class="fa fa-check"></i><b>2.2</b> Peek into Datasets</a></li>
<li class="chapter" data-level="2.3" data-path="load-data.html"><a href="load-data.html#dataset-cleaning"><i class="fa fa-check"></i><b>2.3</b> Dataset Cleaning</a></li>
<li class="chapter" data-level="2.4" data-path="load-data.html"><a href="load-data.html#data-pooling"><i class="fa fa-check"></i><b>2.4</b> Data Pooling</a></li>
<li class="chapter" data-level="2.5" data-path="load-data.html"><a href="load-data.html#batch-correction"><i class="fa fa-check"></i><b>2.5</b> Batch Correction</a></li>
<li class="chapter" data-level="2.6" data-path="load-data.html"><a href="load-data.html#separate-data-set-into-train-and-test-subsets"><i class="fa fa-check"></i><b>2.6</b> Separate Data Set into Train and Test Subsets</a></li>
<li class="chapter" data-level="2.7" data-path="load-data.html"><a href="load-data.html#save-gene-sets"><i class="fa fa-check"></i><b>2.7</b> Save Gene Sets</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="train-models.html"><a href="train-models.html"><i class="fa fa-check"></i><b>3</b> Train Predictive Models</a>
<ul>
<li class="chapter" data-level="3.1" data-path="train-models.html"><a href="train-models.html#glmnet---elastic-net-regularized-generalized-linear-models"><i class="fa fa-check"></i><b>3.1</b> glmnet - Elastic-Net Regularized Generalized Linear Models</a></li>
<li class="chapter" data-level="3.2" data-path="train-models.html"><a href="train-models.html#enet---elastic-net-regularized-generalized-linear-models"><i class="fa fa-check"></i><b>3.2</b> enet - Elastic-Net Regularized Generalized Linear Models</a></li>
<li class="chapter" data-level="3.3" data-path="train-models.html"><a href="train-models.html#knn---k-nearest-neighbors"><i class="fa fa-check"></i><b>3.3</b> knn - k nearest neighbors</a></li>
<li class="chapter" data-level="3.4" data-path="train-models.html"><a href="train-models.html#pam---nearest-shrunken-centroid"><i class="fa fa-check"></i><b>3.4</b> pam - nearest shrunken centroid</a></li>
<li class="chapter" data-level="3.5" data-path="train-models.html"><a href="train-models.html#svmradial---support-vector-machines-rbf-kernel"><i class="fa fa-check"></i><b>3.5</b> svmRadial - Support vector machines (RBF kernel)</a></li>
<li class="chapter" data-level="3.6" data-path="train-models.html"><a href="train-models.html#gbm---boosted-trees"><i class="fa fa-check"></i><b>3.6</b> gbm - Boosted trees</a></li>
<li class="chapter" data-level="3.7" data-path="train-models.html"><a href="train-models.html#xgblinear---extreme-gradient-boosting"><i class="fa fa-check"></i><b>3.7</b> xgbLinear - eXtreme Gradient Boosting</a></li>
<li class="chapter" data-level="3.8" data-path="train-models.html"><a href="train-models.html#rf---random-forests"><i class="fa fa-check"></i><b>3.8</b> rf - Random forests</a></li>
<li class="chapter" data-level="3.9" data-path="train-models.html"><a href="train-models.html#stepldafit---stepwise-linear-discriminant-analysis"><i class="fa fa-check"></i><b>3.9</b> stepLDAFit - Stepwise linear discriminant analysis</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="comp-models.html"><a href="comp-models.html"><i class="fa fa-check"></i><b>4</b> Compare Predictive Models</a>
<ul>
<li class="chapter" data-level="4.1" data-path="comp-models.html"><a href="comp-models.html#load-models"><i class="fa fa-check"></i><b>4.1</b> Load Models</a></li>
<li class="chapter" data-level="4.2" data-path="comp-models.html"><a href="comp-models.html#compare-times"><i class="fa fa-check"></i><b>4.2</b> Compare Times</a></li>
<li class="chapter" data-level="4.3" data-path="comp-models.html"><a href="comp-models.html#compare-prediction-accuracy"><i class="fa fa-check"></i><b>4.3</b> Compare Prediction Accuracy</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="comp-models.html"><a href="comp-models.html#out-of-fold-train-data-accuracy"><i class="fa fa-check"></i><b>4.3.1</b> Out-of-Fold Train Data Accuracy</a></li>
<li class="chapter" data-level="4.3.2" data-path="comp-models.html"><a href="comp-models.html#test-data-accuracy"><i class="fa fa-check"></i><b>4.3.2</b> Test Data Accuracy</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="comp-models.html"><a href="comp-models.html#compare-models-in-terms-of-roc"><i class="fa fa-check"></i><b>4.4</b> Compare Models in Terms of ROC</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="comp-models.html"><a href="comp-models.html#training-data-roc"><i class="fa fa-check"></i><b>4.4.1</b> Training Data ROC</a></li>
<li class="chapter" data-level="4.4.2" data-path="comp-models.html"><a href="comp-models.html#out-of-fold-train-average-over-repeats"><i class="fa fa-check"></i><b>4.4.2</b> Out-of-fold Train: Average over Repeats</a></li>
<li class="chapter" data-level="4.4.3" data-path="comp-models.html"><a href="comp-models.html#compare-predicted-probabilities"><i class="fa fa-check"></i><b>4.4.3</b> Compare Predicted Probabilities</a></li>
<li class="chapter" data-level="4.4.4" data-path="comp-models.html"><a href="comp-models.html#compare-predicted-probabilities---test-set"><i class="fa fa-check"></i><b>4.4.4</b> Compare Predicted Probabilities - Test Set</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="comp-models.html"><a href="comp-models.html#look-at-variable-importance"><i class="fa fa-check"></i><b>4.5</b> Look at Variable Importance</a></li>
<li class="chapter" data-level="4.6" data-path="comp-models.html"><a href="comp-models.html#classification-discussion"><i class="fa fa-check"></i><b>4.6</b> Classification: Discussion</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>5</b> Cluster Analysis</a>
<ul>
<li class="chapter" data-level="5.1" data-path="clustering.html"><a href="clustering.html#bootstrap-aggregation-of-clusters"><i class="fa fa-check"></i><b>5.1</b> Bootstrap Aggregation of Clusters</a></li>
<li class="chapter" data-level="5.2" data-path="clustering.html"><a href="clustering.html#t-sne"><i class="fa fa-check"></i><b>5.2</b> t-SNE</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="clustering.html"><a href="clustering.html#t-sne-pam"><i class="fa fa-check"></i><b>5.2.1</b> t-SNE + PAM</a></li>
<li class="chapter" data-level="5.2.2" data-path="clustering.html"><a href="clustering.html#t-sne-bagclust2"><i class="fa fa-check"></i><b>5.2.2</b> t-SNE + BagClust2</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="clustering.html"><a href="clustering.html#cluster-analysis-discussion"><i class="fa fa-check"></i><b>5.3</b> Cluster Analysis: Discussion</a></li>
<li class="chapter" data-level="5.4" data-path="clustering.html"><a href="clustering.html#next-step-look-for-subgroups-in-the-mss-population---an-open-biological-question."><i class="fa fa-check"></i><b>5.4</b> Next Step: Look for subgroups in the MSS population - an open biological question.</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="summary.html"><a href="summary.html"><i class="fa fa-check"></i><b>6</b> Results Summary</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="LICENSE.md" style="font:em;" target="blank">Copyright (c) 2020 Francois Collin</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Classification and Clustering: Case Study with CRC Gene Expression Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-Guinney:2015aa">
<p>1. Guinney, J., Dienstmann, R., Wang, X., Reyniès, A. de, Schlicker, A., Soneson, C., Marisa, L., Roepman, P., Nyamundanda, G., and Angelino, P. <em>et al.</em> The consensus molecular subtypes of colorectal cancer. Nature medicine <em>21</em>, 1350–1356. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4636487/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4636487/</a>.</p>
</div>
<div id="ref-Boland:1998aa">
<p>2. Boland, C.R., Thibodeau, S.N., Hamilton, S.R., Sidransky, D., Eshleman, J.R., Burt, R.W., Meltzer, S.J., Rodriguez-Bigas, M.A., Fodde, R., and Ranzani, G.N. <em>et al.</em> A national cancer institute workshop on microsatellite instability for cancer detection and familial predisposition: Development of international criteria for the determination of microsatellite instability in colorectal cancer. Cancer Research <em>58</em>, 5248. Available at: <a href="http://cancerres.aacrjournals.org/content/58/22/5248.abstract">http://cancerres.aacrjournals.org/content/58/22/5248.abstract</a>.</p>
</div>
<div id="ref-Lazar:2013aa">
<p>3. Lazar, C., Meganck, S., Taminau, J., Steenhoff, D., Coletta, A., Molter, C., Weiss-Solı́s, D.Y., Duque, R., Bersini, H., and Nowé, A. Batch effect removal methods for microarray gene expression data integration: A survey. Briefings in Bioinformatics <em>14</em>, 469–490. Available at: <a href="http://dx.doi.org/10.1093/bib/bbs037">http://dx.doi.org/10.1093/bib/bbs037</a>.</p>
</div>
<div id="ref-Bolstad:2003aa">
<p>4. Bolstad, B.M., Irizarry, R.A., Åstrand, M., and Speed, T.P. A comparison of normalization methods for high density oligonucleotide array data based on variance and bias. Bioinformatics <em>19</em>, 185–193. Available at: <a href="http://dx.doi.org/10.1093/bioinformatics/19.2.185">http://dx.doi.org/10.1093/bioinformatics/19.2.185</a>.</p>
</div>
<div id="ref-Gagnon-Bartsch:2012aa">
<p>5. Gagnon-Bartsch, J.A., and Speed, T.P. Using control genes to correct for unwanted variation in microarray data. Biostatistics <em>13</em>, 539–552. Available at: <a href="http://dx.doi.org/10.1093/biostatistics/kxr034">http://dx.doi.org/10.1093/biostatistics/kxr034</a>.</p>
</div>
<div id="ref-Molania:2019aa">
<p>6. Molania, R., Gagnon-Bartsch, J.A., Dobrovic, A., and Speed, T.P. A new normalization for nanostring nCounter gene expression data. Nucleic acids research <em>47</em>, 6073–6083. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/31114909">https://pubmed.ncbi.nlm.nih.gov/31114909</a>.</p>
</div>
<div id="ref-Leek:2012aa">
<p>7. Leek, J.T., Johnson, W.E., Parker, H.S., Jaffe, A.E., and Storey, J.D. The sva package for removing batch effects and other unwanted variation in high-throughput experiments. Bioinformatics <em>28</em>, 882–883. Available at: <a href="http://dx.doi.org/10.1093/bioinformatics/bts034">http://dx.doi.org/10.1093/bioinformatics/bts034</a>.</p>
</div>
<div id="ref-Irizarry:2003aa">
<p>8. Irizarry, R.A., Bolstad, B.M., Collin, F., Cope, L.M., Hobbs, B., and Speed, T.P. Summaries of affymetrix genechip probe level data. Nucleic Acids Research <em>31</em>, e15–e15. Available at: <a href="http://dx.doi.org/10.1093/nar/gng015">http://dx.doi.org/10.1093/nar/gng015</a>.</p>
</div>
<div id="ref-Vihinen:2012aa">
<p>9. Vihinen, M. (2012). How to evaluate performance of prediction methods? Measures and their interpretation in variation effect analysis. BMC Genomics <em>13</em>, S2–S2. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3303716/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3303716/</a>.</p>
</div>
<div id="ref-Watanabe:2006aa">
<p>10. Watanabe, T., Kobunai, T., Toda, E., Yamamoto, Y., Kanazawa, T., Kazama, Y., Tanaka, J., Tanaka, T., Konishi, T., and Okayama, Y. <em>et al.</em> Distal colorectal cancers with microsatellite instability (msi) display distinct gene expression profiles that are different from proximal msi cancers. Cancer Research <em>66</em>, 9804. Available at: <a href="http://cancerres.aacrjournals.org/content/66/20/9804.abstract">http://cancerres.aacrjournals.org/content/66/20/9804.abstract</a>.</p>
</div>
<div id="ref-Jorissen:2008aa">
<p>11. Jorissen, R.N., Lipton, L., Gibbs, P., Chapman, M., Desai, J., Jones, I.T., Yeatman, T.J., East, P., Tomlinson, I.P.M., and Verspaget, H.W. <em>et al.</em> DNA copy-number alterations underlie gene expression differences between microsatellite stable and unstable colorectal cancers. Clinical Cancer Research <em>14</em>, 8061. Available at: <a href="http://clincancerres.aacrjournals.org/content/14/24/8061.abstract">http://clincancerres.aacrjournals.org/content/14/24/8061.abstract</a>.</p>
</div>
<div id="ref-Alhopuro:2012aa">
<p>12. Alhopuro, P., Sammalkorpi, H., Niittymäki, I., Biström, M., Raitila, A., Saharinen, J., Nousiainen, K., Lehtonen, H.J., Heliövaara, E., and Puhakka, J. <em>et al.</em> (2012). Candidate driver genes in microsatellite-unstable colorectal cancer. International Journal of Cancer <em>130</em>, 1558–1566. Available at: <a href="http://dx.doi.org/10.1002/ijc.26167">http://dx.doi.org/10.1002/ijc.26167</a>.</p>
</div>
<div id="ref-Watanabe:2012aa">
<p>13. Watanabe, T., Kobunai, T., Yamamoto, Y., Matsuda, K., Ishihara, S., Nozawa, K., Yamada, H., Hayama, T., Inoue, E., and Tamura, J. <em>et al.</em> Chromosomal instability (cin) phenotype, cin high or cin low, predicts survival for colorectal cancer. Journal of Clinical Oncology <em>30</em>, 2256–2264. Available at: <a href="https://doi.org/10.1200/JCO.2011.38.6490">https://doi.org/10.1200/JCO.2011.38.6490</a>.</p>
</div>
<div id="ref-Schlicker:2012aa">
<p>14. Schlicker, A., Beran, G., Chresta, C.M., McWalter, G., Pritchard, A., Weston, S., Runswick, S., Davenport, S., Heathcote, K., and Castro, D.A. <em>et al.</em> (2012). Subtypes of primary colorectal tumors correlate with response to targeted treatment in colorectal cell lines. BMC Medical Genomics <em>5</em>, 66–66. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3543849/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3543849/</a>.</p>
</div>
<div id="ref-Barras:2017aa">
<p>15. Barras, D., Missiaglia, E., Wirapati, P., Sieber, O.M., Jorissen, R.N., Love, C., Molloy, P.L., Jones, I.T., McLaughlin, S., and Gibbs, P. <em>et al.</em> <span class="math inline">\(&lt;\)</span>em<span class="math inline">\(&gt;\)</span>BRAF<span class="math inline">\(&lt;\)</span>/em<span class="math inline">\(&gt;\)</span> <span class="math inline">\(&lt;\)</span>em<span class="math inline">\(&gt;\)</span>V600E<span class="math inline">\(&lt;\)</span>/em<span class="math inline">\(&gt;\)</span> mutant colorectal cancer subtypes based on gene expression. Clinical Cancer Research <em>23</em>, 104. Available at: <a href="http://clincancerres.aacrjournals.org/content/23/1/104.abstract">http://clincancerres.aacrjournals.org/content/23/1/104.abstract</a>.</p>
</div>
<div id="ref-Kirzin:2014aa">
<p>16. Kirzin, S., Marisa, L., Guimbaud, R., De Reynies, A., Legrain, M., Laurent-Puig, P., Cordelier, P., Pradère, B., Bonnet, D., and Meggetto, F. <em>et al.</em> (2014). Sporadic early-onset colorectal cancer is a specific sub-type of cancer: A morphological, molecular and genetics study. PLoS ONE <em>9</em>, e103159. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4118858/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4118858/</a>.</p>
</div>
<div id="ref-Schmit:2015aa">
<p>17. Schmit, S.L., Gollub, J., Shapero, M.H., Huang, S.-C., Rennert, H.S., Finn, A., Rennert, G., and Gruber, S.B. MicroRNA polymorphisms and risk of colorectal cancer. Cancer Epidemiology Biomarkers &amp;amp; Prevention <em>24</em>, 65. Available at: <a href="http://cebp.aacrjournals.org/content/24/1/65.abstract">http://cebp.aacrjournals.org/content/24/1/65.abstract</a>.</p>
</div>
<div id="ref-Yagi:2010aa">
<p>18. Yagi, K., Akagi, K., Hayashi, H., Nagae, G., Tsuji, S., Isagawa, T., Midorikawa, Y., Nishimura, Y., Sakamoto, H., and Seto, Y. <em>et al.</em> Three dna methylation epigenotypes in human colorectal cancer. Clinical Cancer Research <em>16</em>, 21. Available at: <a href="http://clincancerres.aacrjournals.org/content/16/1/21.abstract">http://clincancerres.aacrjournals.org/content/16/1/21.abstract</a>.</p>
</div>
<div id="ref-Laiho:2006aa">
<p>19. Laiho, P., Kokko, A., Vanharanta, S., Salovaara, R., Sammalkorpi, H., Jarvinen, H., Mecklin, J.-P., Karttunen, T.J., Tuppurainen, K., and Davalos, V. <em>et al.</em> Serrated carcinomas form a subclass of colorectal cancer with distinct molecular basis. Oncogene <em>26</em>, 312–320. Available at: <a href="http://dx.doi.org/10.1038/sj.onc.1209778">http://dx.doi.org/10.1038/sj.onc.1209778</a>.</p>
</div>
<div id="ref-Tibshirani:2002ab">
<p>20. Tibshirani, R., Hastie, T., Narasimhan, B., and Chu, G. Diagnosis of multiple cancer types by shrunken centroids of gene expression. Proceedings of the National Academy of Sciences <em>99</em>, 6567. Available at: <a href="http://www.pnas.org/content/99/10/6567.abstract">http://www.pnas.org/content/99/10/6567.abstract</a>.</p>
</div>
<div id="ref-Banerjea:2004aa">
<p>21. Banerjea, A., Ahmed, S., Hands, R.E., Huang, F., Han, X., Shaw, P.M., Feakins, R., Bustin, S.A., and Dorudi, S. (2004). Colorectal cancers with microsatellite instability display mRNA expression signatures characteristic of increased immunogenicity. Molecular Cancer <em>3</em>, 21. Available at: <a href="http://dx.doi.org/10.1186/1476-4598-3-21">http://dx.doi.org/10.1186/1476-4598-3-21</a>.</p>
</div>
<div id="ref-CROCE">
<p>22. Lanza, G., Ferracin, M., Gafà, R., Veronese, A., Spizzo, R., Pichiorri, F., Liu, C.-g., Calin, G.A., Croce, C.M., and Negrini, M. (2007). MRNA/microRNA gene expression profile in microsatellite unstable colorectal cancer. Molecular Cancer <em>6</em>, 54–54. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2048978/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2048978/</a>.</p>
</div>
<div id="ref-Jorissen:2009aa">
<p>23. Jorissen, R.N., Gibbs, P., Christie, M., Prakash, S., Lipton, L., Desai, J., Kerr, D., Aaltonen, L.A., Arango, D., and Kruhøffer, M. <em>et al.</em> Metastasis-associated gene expression changes predict poor outcomes in patients with dukes’stage b and c colorectal cancer. Clinical cancer research : an official journal of the American Association for Cancer Research <em>15</em>, 7642–7651. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2920750/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2920750/</a>.</p>
</div>
<div id="ref-Koinuma:2005aa">
<p>24. Koinuma, K., Yamashita, Y., Liu, W., Hatanaka, H., Kurashina, K., Wada, T., Takada, S., Kaneda, R., Choi, Y.L., and Fujiwara, S.-I. <em>et al.</em> Epigenetic silencing of axin2 in colorectal carcinoma with microsatellite instability. Oncogene <em>25</em>, 139–146. Available at: <a href="http://dx.doi.org/10.1038/sj.onc.1209009">http://dx.doi.org/10.1038/sj.onc.1209009</a>.</p>
</div>
<div id="ref-Kruhoffer:2005aa">
<p>25. Kruhøffer, M., Jensen, J.L., Laiho, P., Dyrskjøt, L., Salovaara, R., Arango, D., Birkenkamp-Demtroder, K., Sørensen, F.B., Christensen, L.L., and Buhl, L. <em>et al.</em> Gene expression signatures for colorectal cancer microsatellite status and hnpcc. British Journal of Cancer <em>92</em>, 2240–2248. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2361815/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2361815/</a>.</p>
</div>
<div id="ref-Mori:2003aa">
<p>26. Mori, Y., Selaru, F.M., Sato, F., Yin, J., Simms, L.A., Xu, Y., Olaru, A., Deacu, E., Wang, S., and Taylor, J.M. <em>et al.</em> The impact of microsatellite instability on the molecular phenotype of colorectal tumors. Cancer Research <em>63</em>, 4577. Available at: <a href="http://cancerres.aacrjournals.org/content/63/15/4577.abstract">http://cancerres.aacrjournals.org/content/63/15/4577.abstract</a>.</p>
</div>
<div id="ref-Haury:2011aa">
<p>27. Haury, A.-C., Gestraud, P., and Vert, J.-P. (2011). The influence of feature selection methods on accuracy, stability and interpretability of molecular signatures. PloS one <em>6</em>, e28210–e28210. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/22205940">https://pubmed.ncbi.nlm.nih.gov/22205940</a>.</p>
</div>
<div id="ref-Guyon:2003aa">
<p>28. Guyon, I., and Elisseeff, A. (2003). An introduction to variable and feature selection. J. Mach. Learn. Res. <em>3</em>, 1157–1182.</p>
</div>
<div id="ref-Ein-Dor:2005aa">
<p>29. Ein-Dor, L., Kela, I., Getz, G., Givol, D., and Domany, E. (2005). Outcome signature genes in breast cancer: Is there a unique set? Bioinformatics <em>21</em>, 171–178. Available at: <a href="https://doi.org/10.1093/bioinformatics/bth469">https://doi.org/10.1093/bioinformatics/bth469</a>.</p>
</div>
<div id="ref-Ein-Dor:2006aa">
<p>30. Ein-Dor, L., Zuk, O., and Domany, E. Thousands of samples are needed to generate a robust gene list for predicting outcome in cancer. Proceedings of the National Academy of Sciences of the United States of America <em>103</em>, 5923–5928. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/16585533">https://pubmed.ncbi.nlm.nih.gov/16585533</a>.</p>
</div>
<div id="ref-Lai:2006aa">
<p>31. Lai, C., Reinders, M.J., Veer, L.J. van’t, and Wessels, L.F. (2006). A comparison of univariate and multivariate gene selection techniques for classification of cancer datasets. BMC Bioinformatics <em>7</em>, 235. Available at: <a href="https://doi.org/10.1186/1471-2105-7-235">https://doi.org/10.1186/1471-2105-7-235</a>.</p>
</div>
<div id="ref-Abeel:2010aa">
<p>32. Abeel, T., Helleputte, T., Van de Peer, Y., Dupont, P., and Saeys, Y. (2010). Robust biomarker identification for cancer diagnosis with ensemble feature selection methods. Bioinformatics <em>26</em>, 392–398. Available at: <a href="https://doi.org/10.1093/bioinformatics/btp630">https://doi.org/10.1093/bioinformatics/btp630</a>.</p>
</div>
<div id="ref-Zou:2005aa">
<p>33. Zou, H., and Hastie, T. Regularization and variable selection via the elastic net. Journal of the Royal Statistical Society: Series B (Statistical Methodology) <em>67</em>, 301–320. Available at: <a href="https://doi.org/10.1111/j.1467-9868.2005.00503.x">https://doi.org/10.1111/j.1467-9868.2005.00503.x</a>.</p>
</div>
<div id="ref-Meinshausen:2010aa">
<p>34. Meinshausen, N., and Bühlmann, P. Stability selection. Journal of the Royal Statistical Society: Series B (Statistical Methodology) <em>72</em>, 417–473. Available at: <a href="https://doi.org/10.1111/j.1467-9868.2010.00740.x">https://doi.org/10.1111/j.1467-9868.2010.00740.x</a>.</p>
</div>
<div id="ref-Verhaak:2010aa">
<p>35. Verhaak, R.G., Hoadley, K.A., Purdom, E., Wang, V., Qi, Y., Wilkerson, M.D., Miller, C.R., Ding, L., Golub, T., and Mesirov, J.P. <em>et al.</em> An integrated genomic analysis identifies clinically relevant subtypes of glioblastoma characterized by abnormalities in pdgfra, idh1, egfr and nf1. Cancer cell <em>17</em>, 98. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2818769/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2818769/</a>.</p>
</div>
<div id="ref-Kaufmann:1990aa">
<p>36. Kaufman, L., and Rousseeuw, P.J. (1990). Finding groups in data: An introduction to cluster analysis (John Wiley).</p>
</div>
<div id="ref-Dudoit:2003aa">
<p>37. Dudoit, S., and Fridlyand, J. Bagging to improve the accuracy of a clustering procedure. Bioinformatics <em>19</em>, 1090–1099. Available at: <a href="http://dx.doi.org/10.1093/bioinformatics/btg038">http://dx.doi.org/10.1093/bioinformatics/btg038</a>.</p>
</div>
<div id="ref-Li:2011aa">
<p>38. Li, H.G., Wu, G.Q., Hu, X.G., Zhang, J., Li, L., and Wu, X. (2011). K-means clustering with bagging and mapreduce. 2011 44th Hawaii International Conference on System Sciences, 1–8.</p>
</div>
</div>

</div>






            </section>

          </div>
        </div>
      </div>
<a href="summary.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["ExploreCRC.pdf", "ExploreCRC.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
